Sparse checkout for git pulls

tetracionist opened the pull request:
Added support for sparse-checkout when cloning a GitHub repo (https://github.com/PrefectHQ/prefect/issues/15185). 
This is ideal for teams that have larger repos who don't want to clone all the folders in their repo. 

When directories are specified in the Prefect.yaml or GitHubRepository storage class, we will use sparse checkout to get the directories and any file in the root directory (cone-mode).

Example usage in a `Prefect.yaml`

```sh
pull:
- prefect.deployments.steps.git_clone:
    repository: https://github.com/tetracionist/prefect.git
    branch: sparse-checkout-for-git-pulls
    access_token:
    directories: [src/integrations/prefect-azure, src/integrations/prefect-dask]

```

### Checklist
<!-- These boxes may be checked after opening the pull request. -->

- [x] This pull request references any related issue by including "closes `<link to issue>`"
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] If this pull request adds new functionality, it includes unit tests that cover the changes
- [ ] If this pull request removes docs files, it includes redirect settings in `mint.json`.
- [x] If this pull request adds functions or classes, it includes helpful docstrings.

desertaxle reviewed (commented):
Thanks for the PR @tetracionist! This is looking pretty good, but I have a couple of questions about this implementation after a first pass.

tetracionist commented on src/prefect/runner/storage.py:
```diff
@@ -242,21 +282,39 @@ async def _clone_repo(self):
-        # Limit git history and set path to clone to
-        cmd += ["--depth", "1", str(self.destination)]
+        # For sparse-checkout it is recommended to use --filter=blob:none to reduce disk-space
+        if self._sparse_checkout_mode:
```
I put this in the else block as I was unsure what submodules did, but are they something to do with adding code from another repo into the cloned repo?

I think this would be easy with cone mode but might be trickier without.
Would it be useful to also sparse-checkout the submodule?

desertaxle replied:
Yeah, submodules allow you to reference another git repository from a parent repository.  I'm not sure how sparse checkout would interact with submodules, but I think it'd be useful to apply sparse checkout to submodules, also.

tetracionist commented on src/prefect/runner/storage.py:
```diff
@@ -178,6 +188,32 @@ def _repository_url_with_credentials(self) -> str:
+        Uses sparse-checkout on repository
+        """
+
+        cmd = ["git", "sparse-checkout", "init"]
```
I think this should be okay, I will try using `git sparse-checkout set`

tetracionist commented:
@desertaxle Thanks for reviewing back in October, I had a chance to have another look at this PR again.

The main changes I have made are as follows:

- Now using `sparse-checkout set` instead of init
- Added an async function to check if repository is in sparse-checkout mode
- Removed cone-mode 
- Two new tests for sparse-checkout
- Can still use submodules if using sparse-checkout
- I ditched the filter blob none and ended up sticking with shallow clones (after reading these [docs](https://github.blog/open-source/git/get-up-to-speed-with-partial-clone-and-shallow-clone/) it seemed to be the most sensible option)


I also used the configuration and flow below to test what folders get cloned and that worked as expected. 
Please would you be able to to review this again?

```yaml
definitions:
  actions:
    pull_staging: &pull_staging
      - prefect.deployments.steps.git_clone:
          repository: https://github.com/mannum/repo.git
          branch: test-sparse
          access_token: '{{ prefect.blocks.secret.my-gh-token }}'
          directories: ["orchestration", "ingestion"]

```

```python
from pathlib import Path

from prefect import flow
from prefect.logging import get_run_logger


@flow(name="dir check flow")
def dir_check_flow():

    logger = get_run_logger()
   
    directory = Path()
    logger.info(f"{list(directory.rglob("*"))}")


if __name__ == "__main__":
    dir_check_flow()

```

desertaxle reviewed (changes requested):
LGTM! Thanks for the awesome contribution @tetracionist!

desertaxle reviewed (approved):
oops clicked the wrong radio button
